data-analytics/iot-nirvana/app-engine/src/main/java/com/google/cloud/demo/iot/nirvana/servlet/rest/GetPostGce.java

```diff
@@ -0,0 +1,157 @@
+    Gson gson = new Gson();
+
+    resp.setContentType(com.google.common.net.MediaType.JSON_UTF_8.toString());
+    restResponse.setCode(javax.servlet.http.HttpServletResponse.SC_OK);
+    restResponse.setMessage("Task enqueued");
+    resp.setStatus(javax.servlet.http.HttpServletResponse.SC_OK);
+    resp.getWriter().println(gson.toJson(restResponse));
+
+    return;
+  }
+
+  public void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException {
+    try {
+      // manage dev env
+      if (!Utils.isGaeProduction()) {
+        resp.addHeader("Access-Control-Allow-Origin", Constants.GAE_DEV_ADDRESS);
+      }
+      // include cookies
+      resp.addHeader("Access-Control-Allow-Methods", "GET, POST, PATCH, PUT, DELETE, OPTIONS");
+      resp.addHeader("Access-Control-Allow-Credentials", "true");
+
+      // create default Google AppEngine credential
+      GoogleCredential credential =
+          GoogleCredential.getApplicationDefault().createScoped(ComputeScopes.all());
+
+      // access to Compute Engine
+      // Create compute engine object for listing instances
+      Compute compute =
+          new Compute.Builder(Constants.HTTP_TRANSPORT, Constants.JSON_FACTORY, null)
+              .setApplicationName(Constants.APPLICATION_NAME)
+              .setHttpRequestInitializer(credential)
+              .build();
+
+      // list instances
+      String nextPageToken = null;
+      int instancesCounter = 0;
+      InstanceList instanceList =
+          compute
+              .instances()
+              .list(Constants.GAE_APP_ID, Constants.GCE_MACHINE_DEFAULT_ZONE)
+              .execute();
+      nextPageToken = instanceList.getNextPageToken();
+      if (instanceList.getItems() != null) {
+        instancesCounter += instanceList.getItems().size();
+      }
+      // if there are more token to elaborate
+      while (nextPageToken != null) {
+        // get instances
+        instanceList =
+            compute
+                .instances()
+                .list(Constants.GAE_APP_ID, Constants.GCE_MACHINE_DEFAULT_ZONE)
+                .setPageToken(nextPageToken)
+                .execute();
+        // save instance information
+        if (instanceList.getItems() != null) {
+          instancesCounter += instanceList.getItems().size();
+        }
+        // update nextPageToken
+        nextPageToken = instanceList.getNextPageToken();
+      }
+
+      // return data
+      RestResponse restResponse = new RestResponse();
+      Gson gson = new Gson();
+
+      resp.setContentType(com.google.common.net.MediaType.JSON_UTF_8.toString());
+      restResponse.setCode(javax.servlet.http.HttpServletResponse.SC_OK);
+      restResponse.setMessage("" + instancesCounter);
+      resp.setStatus(javax.servlet.http.HttpServletResponse.SC_OK);
+      resp.getWriter().println(gson.toJson(restResponse));
+
+      return;
+    } catch (Exception ex) {
+      LOG.warning(Throwables.getStackTraceAsString(ex));
```

Comment:
There are so many things which can go bad here...Returning 500 may make more sense as 400 may confuse developers to look in their Request object which issue would most probably be on server side. Ideally, we should handle different errors ( auth, quota, etc..) differently.

Reply:
As a first step I changed the code to return HTTP code 500 with a message recommending to check the application log for more details. Change done in commit https://github.com/GoogleCloudPlatform/professional-services/pull/61/commits/63823a3d1af592703a4b4930d26f09f9651a26de